File: crates/airc-client/src/event.rs
//! IRC events — typed representations of incoming messages.

use std::time::SystemTime;

use serde::{Deserialize, Serialize};

// Re-export proto types as the canonical message types.
pub use airc_shared::common::ChannelMessage;
pub use airc_shared::common::MessageKind;

/// A high-level IRC event parsed from the wire.
///
/// This is an internal client event type — NOT a protobuf type. It carries
/// richer variant data (Registered, Raw, etc.) that are only meaningful
/// inside the client library.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum IrcEvent {
    /// Successfully registered with the server.
    Registered {
        nick: String,
        server: String,
        message: String,
    },
    /// A message in a channel or private message.
    Message(ChannelMessage),
    /// Someone joined a channel.
    Join { nick: String, channel: String },
    /// Someone left a channel.
    Part {
        nick: String,
        channel: String,
        reason: Option<String>,
    },
    /// Someone quit IRC.
    Quit {
        nick: String,
        reason: Option<String>,
    },
    /// Someone was kicked from a channel.
    Kick {
        channel: String,
        nick: String,
        by: String,
        reason: Option<String>,
    },
    /// Channel topic changed.
    TopicChange {
        channel: String,
        topic: String,
        set_by: String,
    },
    /// Nick change.
    NickChange { old_nick: String, new_nick: String },
    /// A notice (from server or user).
    Notice {
        from: Option<String>,
        target: String,
        text: String,
    },
    /// Connection was lost (will attempt reconnect if auto-reconnect is enabled).
    Disconnected { reason: String },
    /// A reconnection attempt is starting.
    Reconnecting { attempt: u32 },
    /// Successfully reconnected after a disconnection.
    Reconnected,
    /// A single MOTD line from the server (RPL_MOTD 372).
    Motd { line: String },
    /// End of MOTD (RPL_ENDOFMOTD 376).
    MotdEnd,
    /// An unhandled/raw IRC message.
    Raw { line: String },
    /// SASL authentication succeeded.  `account` is the canonical account name
    /// confirmed by the server (from RPL_LOGGEDIN 900).
    SaslLoggedIn { account: String },
    /// SASL authentication failed (904 ERR_SASLFAIL or 906 ERR_SASLABORTED).
    SaslFailed { code: u16, reason: String },
    /// A user's away status changed (IRCv3 away-notify).
    /// `message` is `None` when the user returned from away.
    Away {
        nick: String,
        message: Option<String>,
    },
    /// A user's account changed (IRCv3 account-notify).
    /// `account` is `None` when the user logged out (`ACCOUNT *`).
    AccountNotify {
        nick: String,
        account: Option<String>,
    },
}

/// Create a new `ChannelMessage` with the current timestamp.
pub fn new_channel_message(
    target: String,
    from: String,
    text: String,
    kind: MessageKind,
) -> ChannelMessage {
    new_channel_message_with_ts(target, from, text, kind, None)
}

/// Create a new `ChannelMessage` with an optional timestamp.
///
/// If `timestamp` is `None`, uses the current system time.
/// If `timestamp` is `Some(ts)`, uses that value (from IRCv3 `server-time`).
pub fn new_channel_message_with_ts(
    target: String,
    from: String,
    text: String,
    kind: MessageKind,
    timestamp: Option<u64>,
) -> ChannelMessage {
    let ts = timestamp.unwrap_or_else(|| {
        SystemTime::now()
            .duration_since(SystemTime::UNIX_EPOCH)
            .unwrap_or_default()
            .as_secs()
    });
    ChannelMessage {
        target,
        from,
        text,
        kind: kind as i32,
        timestamp: ts,
    }
}

File: packages/airc-client-ts/src/event.ts
/**
 * IRC events — typed representations of incoming messages.
 *
 * Mirrors the Rust `IrcEvent` enum from `airc-client/src/event.rs`.
 * Uses a discriminated union on the `type` field.
 */

// ---------------------------------------------------------------------------
// Message types (from proto)
// ---------------------------------------------------------------------------

/** Matches the proto `MessageKind` enum. */
export const enum MessageKind {
  Normal = 0,
  Action = 1,
}

/** A chat message (channel or private) — matches proto `ChannelMessage`. */
export interface ChannelMessage {
  /** Channel name or nick (DM recipient). */
  target: string;
  /** Sender nick. */
  from: string;
  /** Message body. */
  text: string;
  /** Message kind. */
  kind: MessageKind;
  /** Unix timestamp (seconds). */
  timestamp: number;
}

/** Summary of a joined channel's state — matches proto `ChannelStatus`. */
export interface ChannelStatus {
  name: string;
  topic?: string;
  members: number;
  totalMessages: number;
  unread: number;
}

// ---------------------------------------------------------------------------
// IrcEvent discriminated union
// ---------------------------------------------------------------------------

export type IrcEvent =
  | { type: "registered"; nick: string; server: string; message: string }
  | { type: "message"; message: ChannelMessage }
  | { type: "join"; nick: string; channel: string }
  | { type: "part"; nick: string; channel: string; reason?: string }
  | { type: "quit"; nick: string; reason?: string }
  | { type: "kick"; channel: string; nick: string; by: string; reason?: string }
  | { type: "topic_change"; channel: string; topic: string; setBy: string }
  | { type: "nick_change"; oldNick: string; newNick: string }
  | { type: "notice"; from?: string; target: string; text: string }
  | { type: "disconnected"; reason: string }
  | { type: "reconnecting"; attempt: number }
  | { type: "reconnected" }
  | { type: "motd"; line: string }
  | { type: "motd_end" }
  | { type: "sasl_logged_in"; account: string }
  | { type: "sasl_failed"; code: number; reason: string }
  /**
   * A user in a shared channel set or cleared their away status (away-notify).
   * `message` is `undefined` when the user returned from away.
   */
  | { type: "away"; nick: string; message: string | undefined }
  /**
   * A user's NickServ account changed while already connected (account-notify).
   * `account` is `undefined` when the user logged out (`ACCOUNT *`).
   */
  | { type: "account_notify"; nick: string; account: string | undefined }
  /**
   * Initial member list for a channel, from 353 RPL_NAMREPLY.
   * `members` includes mode prefixes (@, +, %, &, ~) if present.
   * Emitted once per 353 line; multiple lines may arrive for large channels.
   */
  | { type: "names"; channel: string; members: string[] }
  /**
   * Topic set at join time, from 332 RPL_TOPIC.
   * Distinct from `topic_change` which is the live TOPIC command mid-session.
   */
  | { type: "topic"; channel: string; topic: string }
  | { type: "raw"; line: string };

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Create a new `ChannelMessage` with the current timestamp. */
export function newChannelMessage(
  target: string,
  from: string,
  text: string,
  kind: MessageKind,
): ChannelMessage {
  return {
    target,
    from,
    text,
    kind,
    timestamp: Math.floor(Date.now() / 1000),
  };
}

/**
 * Create a new `ChannelMessage` using a server-supplied timestamp when
 * available, falling back to the current system time.
 *
 * `serverTs` should be an ISO 8601 string from the `@time=` tag, or
 * `null`/`undefined` if no tag was present.
 */
export function newChannelMessageWithTs(
  target: string,
  from: string,
  text: string,
  kind: MessageKind,
  serverTs: string | null | undefined,
): ChannelMessage {
  let timestamp = Math.floor(Date.now() / 1000);
  if (serverTs) {
    const parsed = Date.parse(serverTs);
    if (!isNaN(parsed)) {
      timestamp = Math.floor(parsed / 1000);
    }
  }
  return { target, from, text, kind, timestamp };
}
